### PR TITLE
s2s config per host type

### PR DESCRIPTION
**Pull request by chrzaszcz**
The goal is to put all s2s options in a map and allow specifying them globally or per host type. The `host_config.s2s` section completely overrides the global `s2s` section now, as this is more straightforward and allows resetting optional values.

Main changes:
- Handling of incoming connections is reorganized to get host type as early as possible.
- As a side effect, more errors are now properly caught and reported. Tests for them are added.
- Coverage of s2s logic is improved.
- Defaults are used for all options that have defaults in the doc.

Smaller changes
- Removed `s2s.address` default as it had no benefit.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/103720) / small_tests / 4b14c9c8344d84baca3598d82db0c6512761ec43
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3523/103720/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3523/103720/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-01-31_07.57.46/lib.mongooseim.logs/run.2022-01-31_07.57.47/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/104014) / small_tests / 33d967f7dc258d5f7e5677aaad647712fc2bb8cb
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3523/104014/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3523/104014/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-01-31_16.33.50/lib.mongooseim.logs/run.2022-01-31_16.33.51/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/104038) / small_tests / f92b7531451fe438236a06266552919b635a6b03
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3523/104038/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3523/104038/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-01-31_17.20.55/lib.mongooseim.logs/run.2022-01-31_17.20.55/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/104094) / small_tests / 0382fbe05f80885ca86b3204360440f2c0c0842d
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3523/104094/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3523/104094/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-02-01_09.18.37/lib.mongooseim.logs/run.2022-02-01_09.18.37/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/104141) / small_tests / bf1ba556a68fb93d495a6fb1429299c5cfbbb620
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3523/104141/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3523/104141/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-02-01_11.56.30/lib.mongooseim.logs/run.2022-02-01_11.56.31/suite.log.html)

**Comment by mongoose-im** [comment]
[internal_mnesia_24](https://circleci.com/gh/esl/MongooseIM/104412) / internal_mnesia / 3a24ffd509140025e499db098e6d2fc54af48932
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3523/104412/internal_mnesia.24.0.5)

**Comment by mongoose-im** [comment]
[internal_mnesia_24](https://circleci.com/gh/esl/MongooseIM/104438) / internal_mnesia / 77126ba24d24e6705d132f548b296712efa949b2
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3523/104438/internal_mnesia.24.0.5)

**Comment by mongoose-im** [comment]
[internal_mnesia_24](https://circleci.com/gh/esl/MongooseIM/104461) / internal_mnesia / c70b4d6d0ace4f47af8144db29dcb35ee54c59dc
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3523/104461/internal_mnesia.24.0.5)

**Comment by mongoose-im** [comment]
[small_tests_23](https://circleci.com/gh/esl/MongooseIM/104600) / small_tests / 6babcd73bcf3f8bb13be361c92558f2aa06dea77
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3523/104600/small_tests.23.1) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3523/104600/small_tests.23.1/small/ct_run.mongooseim@localhost.2022-02-02_14.39.28/lib.mongooseim.logs/run.2022-02-02_14.39.29/suite.log.html)

**Review by Premwoik — APPROVED** [review]
Looks good to me :D

File: src/ejabberd_s2s.erl [hunk 1]
```diff
@@ -601,18 +591,33 @@ get_s2s_state(S2sPid)->
...
                  {ok, SecretFromConfig} ->
                      SecretFromConfig;
                  {error, not_found} ->
                      base16:encode(crypto:strong_rand_bytes(10))
              end,
-    mnesia:transaction(fun() -> mnesia:write(#s2s_shared{secret = Secret}) end).
+    mnesia:write(#s2s_shared{host_type = HostType, secret = Secret}).
+
+-spec lookup_certfile(mongooseim:host_type()) -> {ok, string()} | {error, not_found}.
+lookup_certfile(HostType) ->
```
[the remark this answers is not in the document]
Comment (chrzaszcz) [reply src/ejabberd_s2s.erl:617]: I wouldn't say `domain_certfile` is global - it is per host type, but it is in the general section.

According to the docs for `domain_certfile`: 
> This option overrides the configured certificate file for specific local XMPP domains. This option applies to S2S and C2S connections.

It took precedence over s2s and c2s settings before and I didn't want to change that. The only difference is that you can set `s2s.certfile` per domain now, but usually you would set it globally, and there is no way to tell how it was set.

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/104689) / small_tests / db060abeec62877f47edd1fde8793816420d70e2
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3523/104689/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3523/104689/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-02-03_07.05.33/lib.mongooseim.logs/run.2022-02-03_07.05.33/suite.log.html)

**Pull request merged**